src/Pirouette/SMT/Constraints.hs

```diff
@@ -1,167 +1,276 @@
+      | otherwise = Just $ unifyNewMetaWith cs v u
+
+    -- Like 'unifyMetaWith', but assumes that the 'meta' in question is not
+    -- known by our current constraints.
+    unifyNewMetaWith ::
+      ConstraintSet lang meta ->
+      meta ->
+      TermMeta lang meta ->
+      ConstraintSet lang meta
+    unifyNewMetaWith cs@ConstraintSet {..} v (SystF.App (SystF.Meta u) [])
+      -- Are v and u the same? Great, nothing to do
+      | v == u = cs
+      -- Is u actually equivalent to anything smaller than u? Try that instead!
+      | Just u' <- M.lookup u csMetaEq = unifyNewMetaWith cs v (SystF.App (SystF.Meta u') [])
+      -- None of the above? Register the equivalence of v and u while altering the
+      -- assignments (if any and if necessary)
+      | otherwise =
+        let newMetaEq = M.insert (max v u) (min v u) csMetaEq
+            -- If u already had a term assigned to it, and v is smaller, we need to
+            -- update the assignments too!
+            newAssignments = case M.lookup u csAssignments of
+              Just tu | v < u -> M.insert v tu (M.delete u csAssignments)
+              _ -> csAssignments
+         in cs {csMetaEq = newMetaEq, csAssignments = newAssignments}
+    unifyNewMetaWith cs v u =
+      cs {csAssignments = M.insert v u (csAssignments cs)}
+
+    unifyArgWith ::
+      ConstraintSet lang meta ->
+      ArgMeta lang meta ->
+      ArgMeta lang meta ->
+      Maybe (ConstraintSet lang meta)
+    unifyArgWith cs (SystF.TermArg x) (SystF.TermArg y) = unifyWith cs x y
+    unifyArgWith cs (SystF.TyArg _) (SystF.TyArg _) = Just cs -- TODO: unify types too? I don't think so
+    unifyArgWith _ _ _ = Nothing
+
+expandDefOf :: (Ord meta) => ConstraintSet lang meta -> meta -> Maybe (TermMeta lang meta)
+expandDefOf cs v =
+  M.lookup v (csAssignments cs)
+    <|> (M.lookup v (csMetaEq cs) >>= expandDefOf cs)
+
+-- | Since the translation of individual constraints can fail,
 -- the translation of constraints does not always carry all the information it could.
 -- So the boolean indicates if every atomic constraint have been translated.
 -- A 'False' indicates that some have been forgotten during the translation.
-constraintToSExpr ::
+constraintSetToSExpr ::
+  forall lang meta m.
   (LanguageSMT lang, ToSMT meta, PirouetteReadDefs lang m) =>
   S.Set Name ->
-  Constraint lang meta ->
+  ConstraintSet lang meta ->
   m (Bool, UsedAnyUFs, PureSMT.SExpr)
-constraintToSExpr knownNames (And constraints) = do
-  atomTrads <- mapM (runTranslator . atomicConstraintToSExpr knownNames) constraints
-  let (translations, usedUFs) = unzip (rights atomTrads)
-  return (all isRight atomTrads, mconcat usedUFs, PureSMT.andMany translations)
-constraintToSExpr _ Bot = return (True, NotUsedAnyUFs, PureSMT.bool False)
+constraintSetToSExpr knownNames ConstraintSet {..} = do
+  eAssignments <- mapM (runTranslator . uncurry trAssignment) $ M.toList csAssignments
+  eNotEq <- mapM (runTranslator . trRelations) csRelations
```

Comment:
This is related to my other comment, but why is this called `eNotEq`? Equality also seems to be translated